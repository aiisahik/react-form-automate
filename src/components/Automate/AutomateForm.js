import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import AutomateField from './AutomateField';

const propTypes = {
    triggerSubmit: PropTypes.func,
    submitting: PropTypes.bool,
    renderMap: PropTypes.object,
    data: PropTypes.array,
    validatorMap: PropTypes.object,
}

const defaultProps = {
    triggerSubmit: () => {},
    submitting: false,
    renderMap: {},
    data: [],
    validatorMap: {},
};

class AutomateForm extends PureComponent {
    constructor(props) {
        super(props);
        this.fields = {};
        const validations = this.props.data.reduce((obj, field) => (
            Object.assign(obj, { [field.fieldName]: field.validations })
        ), {});
        this.state = {
            formData: {},
            validity: {},
            validations: validations,
            submitting: this.props.submitting || false,
            submitted: false,
        };
        this.onInputChange = this.onInputChange.bind(this);
        this.componentDidMount = this.componentDidMount.bind(this);
        this.renderField = this.renderField.bind(this);
        this.submit = this.submit.bind(this);
    }
    componentDidMount() {
        this.props.triggerSubmit(this.submit)
    }   
    componentWillReceiveProps(newProps){
        if (newProps.submitting === true && !this.state.submitted){
            this.setState({
                submitted: newProps.submitting
            });
        }
    }
    submit(){
        const self = this;
        return new Promise((resolve, reject) => {
            this.isValidForm().then(isFormValid => {
                if (isFormValid){
                    resolve(self.props.submit(self.state.formData));
                } else {
                    const failedFields = []
                    Object.keys(self.state.validity).map((fieldName) => {
                        if (self.state.validity[fieldName] === false){
                            failedFields.push(fieldName);
                        }
                    });
                    reject(null);
                }
            })
        });
    }
    isValidForm(){
        const self = this;
        const validationPromises = Object.keys(this.fields).map((fieldName) => {
            if (self.fields[fieldName] && typeof self.fields[fieldName].validate === 'function'){
                return self.fields[fieldName].validate();
                // // update our state.validity object 
                // self.state.validity[fieldName] = fieldValidity;
                // return fieldValidity;
            } else {
                //weird result?
                return Promise.resolve(false);;
            }
        });
        return new Promise((resolve, reject) => {
            Promise.all(validationPromises).then(fieldsValidity => {
                resolve(!Object.values(fieldsValidity).includes(false));
            });
        });
    }
    onInputChange(fieldName, val, valid, formData){
        if (!formData){
            formData = {};
        }
        this.setState(Object.assign({}, this.state, {
            formData: Object.assign({}, this.state.formData, formData, {
                [fieldName]: val
            }),
            validity: typeof valid === 'boolean' ? Object.assign({}, this.state.validity, {
                [fieldName]: valid
            }) : this.state.validity,
        }));

    }
    render() {
        const self = this;
        return (
            <form className="form" noValidate>
                {this.props.data.map((fieldData, index) => (
                    self.renderField(fieldData, index)
                ))}
            </form>
        );
    }
    renderField(fieldData, index) {
        fieldData.render = this.props.renderMap[fieldData.type];
        return (
            <AutomateField
                {...fieldData}
                ref={(fieldRef) => { this.fields[fieldData.fieldName] = fieldRef; }}
                key={index}
                onChange={this.onInputChange}
                submitted={this.state.submitted}
                validatorMap={this.props.validatorMap}
            />
        );
    }
}

AutomateForm.propTypes = propTypes;
AutomateForm.defaultProps = defaultProps;

export default AutomateForm;
