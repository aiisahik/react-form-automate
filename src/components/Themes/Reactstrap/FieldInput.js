import React from 'react';
import classNames from 'classnames';
import { FormGroup, Input, Label } from 'reactstrap';
import FieldValidationMessages from './FieldValidationMessages';

const FieldInput = (props) => {
    return (
        <FormGroup
            className={classNames({
                "has-success": props.valid === true,
                "has-danger": props.valid === false,
            })}
        >
            <Label for="username">{props.label}</Label>
            <Input
                name={props.fieldName}
                defaultValue={props.defaultValue}
                type={props.type}
                className={classNames(props.className, {
                    "is-invalid": props.valid === false,
                    "is-valid": props.valid === true,
                })}
                onBlur={props.onBlur}
                onChange={(e) => props.onChange(e)}
                onKeyDown={props.onKeyDown}
                required={props.isRequired}
                maxLength="60"
                placeholder={props.placeholder} />

            <FieldValidationMessages 
                validations={props.validations}
            />
        </FormGroup>
    );
};

export default FieldInput;