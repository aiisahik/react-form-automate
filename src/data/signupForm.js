import React from 'react';

const signupForm = [
    {
        type: 'input',
        fieldName: 'username',
        label: 'Email',
        validations: {
            'email': { message: 'Please provide a real email address' },
            'required': { message: 'Email is required' },
            'minlength': { message: 'Email needs to be at least 10 characters', minlength: 10 },
            'maxlength': { message: 'Email can be a max of 99 characters', maxlength: 99 },
        },
    },
    {
        type: 'input',
        fieldName: 'companyName',
        label: 'Company Name',
        validations: {
            'required': { message: 'Company is required' },
        },
    },
    {
        type: 'input',
        fieldName: 'zipcode',
        label: 'Zip Code',
        validations: {
            'required': { message: 'Zip code is required' },
            'zipcode': { message: 'Please enter a valid US / Canada zipcode' },
        },
    },
    {
        type: 'select',
        fieldName: 'business_size',
        label: 'Business Size',
        defaultValue: 'small',
        options: [
            { value: '', label: '' },
            { value: 'small', label: 'Small (fewer than 25 employees)' },
            { value: 'medium', label: 'Medium (25 to 100 employees)' },
            { value: 'large', label: 'Large (more than 100 employees)' },
        ],
        validations: {
            'required': { message: 'Business Size is required' },
        },
    },
    {
        type: 'checkbox',
        fieldName: 'tos',
        html: 'I accept the Kinnek <a href="/terms/" target="_blank"> Terms of Use</a > and <a href="/privacy/" target="_blank"> Privacy Policy</a>',
        validations: {
            'required': { message: 'Acceptance of the terms is required' },
        },
    },
];

export default signupForm;