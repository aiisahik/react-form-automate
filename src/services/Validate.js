import axios from 'axios';

export const validate = (val, validations) => {
    const validationPromises = Object.keys(validations).map((type) => {
        switch (type) {
            case "required":
                return Promise.resolve({
                    type: type,
                    valid: (typeof val === 'string' && val.length > 0)
                });
            case "minlength":
                if (typeof validations[type].minlength === 'number') {
                    return Promise.resolve({
                        type: type,
                        valid: val && val.length > validations[type].minlength
                    });
                } else {
                    return { valid: {} };
                }
            case "email":
                const reEmail = /[A-Z0-9._%+-]+@[A-Z0-9.-]+.[A-Z]{2,4}/igm;
                return Promise.resolve({
                    type: type,
                    valid: reEmail.test(val)
                });
            case "phone":
                const rePhone = /[0-9\- #()]{10,}/igm;
                return Promise.resolve({
                    type: type,
                    valid: rePhone.test(val)
                });
            case "zipcode":
                return validateZipcode(val);
            default: 
                return Promise.resolve({
                    type: type,
                    valid: false,
                });
        }
    });

    return new Promise((resolve, reject) => {
        Promise.all(validationPromises).then(values => {
            const newValidations = values.reduce((reducedValidations, validation) => {
                if (validation) {
                    return Object.assign(reducedValidations, {
                        [validation.type]: Object.assign(reducedValidations[validation.type],
                            { valid: validation.valid }
                        )
                    });
                } else {
                    return reducedValidations;
                }
            }, validations);
            const newFormData = values.reduce((reducedFormData, validation) => {
                if (validation && validation.formData) {
                    return Object.assign(reducedFormData, validation.formData)
                } else {
                    return reducedFormData;
                }
            }, {});
            resolve({
                validations: newValidations,
                formData: newFormData,
                valid: Object.values(newValidations).every(x => x.valid !== false)
            });
        }).catch(reason => {
            console.log('error', reason);
            reject(false);
        });
    });
}

const validateZipcode = (zipcode) => {
    return new Promise((resolve, reject) => {
        axios.get(`https://maps.googleapis.com/maps/api/geocode/json?address=${zipcode}`)
            .then((result) => {
                console.log("google maps api result", result);
                const locationData = {};
                if (result.data.results.length && result.data.results[0] && result.data.results[0]['address_components'].length){
                    const firstResult = result.data.results[0];
                    const addressComponents = firstResult['address_components'];
                    locationData.country = addressComponents[addressComponents.length - 1]['short_name'];
                    locationData.state = addressComponents[addressComponents.length - 2]['short_name'];
                    if (firstResult.types && firstResult.types.indexOf('postal_code') >= 0) {
                        locationData.city = firstResult['formatted_address'].split(',')[0];
                    }
                    if (locationData.country && 
                        ["CA", "US",].indexOf(locationData.country) >= 0) {
                        resolve({
                            type: 'zipcode',
                            valid: true,
                            formData: locationData
                        });
                    } else {
                        resolve({ type: 'zipcode', valid: false });
                    }
                } else {
                    resolve({ type: 'zipcode', valid: false });
                }
            }).catch((result) => {
                resolve({ type: 'zipcode', valid: false });
            });
    });
}