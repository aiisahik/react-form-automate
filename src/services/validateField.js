// for 
//      a given value, 
//      a set of validations to apply 
//      a mapping of validation type to validator functions
// return an object with: 
//      the original set of validations, each with a key: valid to denote validity under that validation 
//      any additional form data necessary 
//      a boolean value 'valid' denoting if all validation rules had passed 

const validateField = (val, validations, validators) => {
    const validationPromises = Object.keys(validations).map((type) => {
        const validatorFunction = validators[type];
        if (typeof validatorFunction === 'function'){
            return validatorFunction(val, validations[type]);
        } else {
            return Promise.resolve({
                type: type,
                valid: false,
            });
        }
    });
    return new Promise((resolve, reject) => {
        Promise.all(validationPromises).then(values => {
            const newValidations = values.reduce((reducedValidations, validation) => {
                if (validation) {
                    return Object.assign(reducedValidations, {
                        [validation.type]: Object.assign(reducedValidations[validation.type],
                            { valid: validation.valid }
                        )
                    });
                } else {
                    return reducedValidations;
                }
            }, validations);
            const newFormData = values.reduce((reducedFormData, validation) => {
                if (validation && validation.formData) {
                    return Object.assign(reducedFormData, validation.formData)
                } else {
                    return reducedFormData;
                }
            }, {});
            resolve({
                validations: newValidations,
                formData: newFormData,
                valid: Object.values(newValidations).every(x => x.valid !== false)
            });
        }).catch(reason => {
            console.log('error', reason);
            reject(false);
        });
    });
};

export default validateField;
