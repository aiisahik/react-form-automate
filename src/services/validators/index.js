import validateZipcode from './zipcode';

const validateRequired = (val, data) => {
    return Promise.resolve({
        type: "required",
        valid: (typeof val === 'string' && val.length > 0)
    });
}
const validateMinlength = (val, data) => {
    if (typeof data.minlength === 'number') {
        return Promise.resolve({
            type: 'minlength',
            valid: val && val.length >= data.minlength
        });
    } else {
        return { valid: {} };
    }
}
const validateMaxlength = (val, data) => {
    if (typeof data.maxlength === 'number') {
        return Promise.resolve({
            type: 'maxlength',
            valid: val && val.length <= data.maxlength
        });
    } else {
        return { valid: {} };
    }
}
const validateEmail = (val, data) => {
    const reEmail = /[A-Z0-9._%+-]+@[A-Z0-9.-]+.[A-Z]{2,4}/igm;
    return Promise.resolve({
        type: 'email',
        valid: reEmail.test(val)
    });
}

const validatePhone = (val, data) => {
    const rePhone = /[0-9\- #()]{10,}/igm;
    return Promise.resolve({
        type: 'phone',
        valid: rePhone.test(val)
    });
}

const ValidatorMap = {
    'required': validateRequired,
    'minlength': validateMinlength,
    'maxlength': validateMaxlength,
    'email': validateEmail,
    'phone': validatePhone,
    'zipcode': validateZipcode,
};

export {
    validateRequired, 
    validateMinlength, 
    validateMaxlength,
    validateEmail, 
    validatePhone,
    validateZipcode,
    ValidatorMap,
};
