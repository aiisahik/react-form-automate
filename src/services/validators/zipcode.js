import axios from 'axios';

const validateZipcode = (zipcode) => {
    return new Promise((resolve, reject) => {
        axios.get(`https://maps.googleapis.com/maps/api/geocode/json?address=${zipcode}`)
            .then((result) => {
                const locationData = {};
                if (result.data.results.length && result.data.results[0] && result.data.results[0]['address_components'].length) {
                    const firstResult = result.data.results[0];
                    const addressComponents = firstResult['address_components'];
                    locationData.country = addressComponents[addressComponents.length - 1]['short_name'];
                    locationData.state = addressComponents[addressComponents.length - 2]['short_name'];
                    if (firstResult.types && firstResult.types.indexOf('postal_code') >= 0) {
                        locationData.city = firstResult['formatted_address'].split(',')[0];
                    }
                    if (locationData.country &&
                        ["CA", "US",].indexOf(locationData.country) >= 0) {
                        resolve({
                            type: 'zipcode',
                            valid: true,
                            formData: locationData
                        });
                    } else {
                        resolve({ type: 'zipcode', valid: false });
                    }
                } else {
                    resolve({ type: 'zipcode', valid: false });
                }
            }).catch((result) => {
                resolve({ type: 'zipcode', valid: false });
            });
    });
}
export default validateZipcode;