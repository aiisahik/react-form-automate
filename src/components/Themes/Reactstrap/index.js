import FieldCheckbox from './FieldCheckbox';
import FieldInput from './FieldInput';
import FieldSelect from './FieldSelect';
import FieldValidationMessages from './FieldValidationMessages';

const ReactstrapTheme = {
    'input': FieldInput,
    'checkbox': FieldCheckbox,
    'select': FieldSelect,
    'validationMessages': FieldValidationMessages,
};

export {
    ReactstrapTheme,
    FieldCheckbox,
    FieldInput,
    FieldSelect,
    FieldValidationMessages,
};