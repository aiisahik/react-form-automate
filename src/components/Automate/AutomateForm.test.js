import React from 'react';
import AutomateForm from './AutomateForm';
import renderer from 'react-test-renderer';
import signupForm from '../../data/signupForm';
import { ValidatorMap } from '../../services/validators';
import { ReactstrapTheme } from '../Themes/Reactstrap';

test('Renders signup form', () => {
    const component = renderer.create(
        <AutomateForm
            data={signupForm}
            theme={ReactstrapTheme}
            validatorMap={ValidatorMap}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    // // manually trigger the callback
    // tree.props.onMouseEnter();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();

    // // manually trigger the callback
    // tree.props.onMouseLeave();
    // // re-rendering
    // tree = component.toJSON();
    // expect(tree).toMatchSnapshot();
});