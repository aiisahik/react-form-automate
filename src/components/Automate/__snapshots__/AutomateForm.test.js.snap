// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders signup form 1`] = `
<form
  className="form"
  noValidate={true}
>
  <div
    className="form-group"
  >
    <label
      className=" form-control-label"
      htmlFor="username"
    >
      Email
    </label>
    <input
      className="form-control"
      defaultValue={undefined}
      maxLength="60"
      name="username"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      required={undefined}
      type="input"
    />
    <span />
  </div>
  <div
    className="form-group"
  >
    <label
      className=" form-control-label"
      htmlFor="username"
    >
      Company Name
    </label>
    <input
      className="form-control"
      defaultValue={undefined}
      maxLength="60"
      name="companyName"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      required={undefined}
      type="input"
    />
    <span />
  </div>
  <div
    className="form-group"
  >
    <label
      className=" form-control-label"
      htmlFor="username"
    >
      Zip Code
    </label>
    <input
      className="form-control"
      defaultValue={undefined}
      maxLength="60"
      name="zipcode"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder=""
      required={undefined}
      type="input"
    />
    <span />
  </div>
  <div
    className="form-group"
  >
    <label
      className=" form-control-label"
      htmlFor="username"
    >
      Business Size
    </label>
    <select
      className="form-control"
      defaultValue="small"
      name="business_size"
      onChange={[Function]}
      required={undefined}
    >
      <option
        value=""
      >
        
      </option>
      <option
        value="small"
      >
        Small (fewer than 25 employees)
      </option>
      <option
        value="medium"
      >
        Medium (25 to 100 employees)
      </option>
      <option
        value="large"
      >
        Large (more than 100 employees)
      </option>
    </select>
    <span />
  </div>
  <div
    className="form-check"
  >
    <label
      className="custom-control custom-checkbox form-check-label "
      htmlFor={undefined}
    >
      <input
        className="form-check-input"
        defaultValue={undefined}
        name="tos"
        onChange={[Function]}
        type="checkbox"
        value={true}
      />
      <span
        className="custom-control-indicator"
      />
      <div
        className="custom-control-description mb-3"
        dangerouslySetInnerHTML={
          Object {
            "__html": "I accept the Kinnek <a href=\\"/terms/\\" target=\\"_blank\\"> Terms of Use</a > and <a href=\\"/privacy/\\" target=\\"_blank\\"> Privacy Policy</a>",
          }
        }
      />
      <span />
    </label>
  </div>
</form>
`;
